### PR TITLE
Adapt to level config to match other Jupyter repos

### PR DESCRIPTION
**Pull request by blink1073**
- Adopts the `.pre-commit-config.yaml` hooks used by `jupyter_client`, including the Style & static checks from the [Scientific Python Development Guide](https://learn.scientific-python.org/development/guides/style/).
- Switches to `hatch` for build backend and task runner.
- Adds contributing document.
- Removes use of deprecated `.traverse()` in favor of `.findall()`.

### CODE REVIEW

File: tests/test_execute.py
Comment (akhmerov) [inline tests/test_execute.py:549]: `iter` is unnecessary here, right?

File: jupyter_sphinx/execute.py
Comment (akhmerov) [inline jupyter_sphinx/execute.py:126]: Here and later converting to list looks OK, but I wonder if keeping an iterator is cleaner
```suggestion
        if not next(doctree.findall(JupyterCellNode), False):
```

**Review by akhmerov — APPROVED** [review]

File: tests/test_execute.py
Comment (blink1073) [reply tests/test_execute.py]: Fixed

File: jupyter_sphinx/execute.py
Comment (blink1073) [reply jupyter_sphinx/execute.py]: Updated

**Review by akhmerov — APPROVED** [review]

File: .pre-commit-config.yaml
Comment (12rambau) [inline .pre-commit-config.yaml:9]: I usually use less pre-commit hooks relying mostly on the combo: black, prettier, ruff, codespell. Is there any advantage of using all these ones ? 
Is it simply mirroring the good practices of Jupyter packages (in this case I'll keep it as it is) ?

File: pyproject.toml
Comment (12rambau) [inline pyproject.toml:28]: why setting up a "source code" page if it's the same as the "HomePage" ? Is it compulsory ?

File: .pre-commit-config.yaml
Comment (blink1073) [reply .pre-commit-config.yaml:9]: Yes, we've been copy-pasting pre-commit files for a while now.  :smile:

File: pyproject.toml
Comment (blink1073) [reply pyproject.toml:28]: These links were directly pulled from `setup.py`, I'll change the HomePage to `jupyter.org`.

**Comment by 12rambau** [comment]
I was looking at the repository history, and it seems normal "merge" has been used so far. For this PR only "squash and merge" is available, is it also something you changed @blink1073 ? 

Note that if it's something we can change I am more in favor of a normal merge as it gives more credit to big contributions to the repository than small fixes and as long as the naming convention is correctly managed, it still allow the use of a meaningful changelog (the one generated from Github works just fine)

**Comment by 12rambau** [comment]
@blink1073 I have reactivated the regular merge in the settings but I cannot set it back for this specific branch. Is it ok for you if I squash and merge so we can start from here for further development ?

**Comment by blink1073** [comment]
I disabled "Allow merge commits" because they clutter up the history with merges from "main".  Squash merging should be preferred IMO, unless there truly is some reason each commit in a PR has different meaning (which to me means that the PR should have been split).  However, "Rebase merging" can be used in that scenario.

**Comment by 12rambau** [comment]
@blink1073 I got an upvote from @akhmerov so I think this matter could be discussed. On my side I still see a huge disadvantage to squash merging: It gives the same recognition to a typo PR and a huge feature PR. As an example, I was dead inside when I realized that these 2 PRs (https://github.com/pydata/pydata-sphinx-theme/pull/540, https://github.com/pydata/pydata-sphinx-theme/pull/962/files) were given the same importance in my contributing section. 

The problem you mention seems to be an egg and chicken issue as the "merge from main" conflict only occurs when you squash the previous PR. If not, git is very good at reconciling everything and Github provides a very convinient "sync" button that does everything for you from any branch. 

Another example, I tried to contribute to a repository where people were less available for reviews (https://github.com/widgetti/ipyvuetify/pulls) but still using the same convention as the one you suggested: small single-feature PR. The consequence is that I have PR sitting there for 6 months preventing me from moving forward as I need them pushed sequentially. As this repository doesn't get as much maintainer love as the other Jupyter repos, I think it's very relevant to allow some motivated contributors to make big PR without squashing them.

**Comment by blink1073** [comment]
I'm afk this week, and will defer to whatever you both think is best for this repo. My main point is that you can still see all the commits with the rebase option, but they are at least grouped.  As long as the no-op "merge" commits aren't in the history I think it is fine.

**Comment by 12rambau** [comment]
No worries, I'll pile some small PR in the meantime and wait for your return for the bigger ones. 
I tried to do it myself but I don't have access to branch policies and it seems the blocking parameter is coming from here.

**Pull request merged**
